### PR TITLE
Test access points

### PR DESCRIPTION
**Pull request by sauraank**
## Description of change
Added accesspoint tests. Test and document the behaviour of accesspoints with mountpoint.

Refactored tests for accesspoints. They are still under 'access_points_tests' as we dont have accesspoint resource in our CI.

Relevant issues: 
#4 

## Does this change impact existing behavior?
No

No breaking change.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

File: mountpoint-s3/src/metrics/data.rs [hunk 1]
```diff
@@ -50,13 +50,7 @@ impl Metrics {
...
                 )
             };
-            info!(
-                target: METRICS_TARGET_NAME,
-                "{}{}: {}",
-                key.name(),
-                labels,
-                metric,
-            );
+            info!(target: METRICS_TARGET_NAME, "{}{}: {}", key.name(), labels, metric,);
```
Comment (sauraank) [inline mountpoint-s3/src/metrics/data.rs:53]: Changes in `data.rs` and `fs.rs` are due to `make fmt`. Not sure if I should keep it?

File: mountpoint-s3-client/tests/common/mod.rs [hunk 2]
```diff
@@ -17,6 +17,22 @@ fn init_tracing_subscriber() {
     let _ = tracing_subscriber::fmt::try_init();
 }
 
+pub enum AccessPointType {
+    SingleRegion,
+    ObjectLambda,
+    MultiRegion,
+}
+
+pub fn get_unique_test_prefix(test_name: &str) -> String {
```
Comment (passaro) [inline mountpoint-s3-client/tests/common/mod.rs:26]: Can you call this in `get_test_bucket_and_prefix()`?

File: mountpoint-s3-client/tests/common/mod.rs [hunk 3]
```diff
@@ -37,6 +53,35 @@ pub fn get_test_bucket_and_prefix(test_name: &str) -> (String, String) {
     (bucket, prefix)
 }
 
+pub fn get_test_bucket() -> String {
```
Comment (passaro) [inline mountpoint-s3-client/tests/common/mod.rs:56]: Call this in `get_test_bucket_and_prefix()`

File: mountpoint-s3-client/tests/common/mod.rs [hunk 4]
```diff
@@ -37,6 +53,35 @@ pub fn get_test_bucket_and_prefix(test_name: &str) -> (String, String) {
...
+        }
+        AccessPointType::ObjectLambda => {
+            if arn {
+                std::env::var("S3_OLAP_ARN").expect("Set S3_OLAP_ARN to run integration tests")
+            } else {
+                std::env::var("S3_OLAP_ALIAS").expect("Set S3_OLAP_ALIAS to run integration tests")
+            }
+        }
+        AccessPointType::MultiRegion => {
+            // Multi region accesspoints should only be accessed using their ARN
```
Comment (passaro) [inline mountpoint-s3-client/tests/common/mod.rs:77]: Add an assert

File: mountpoint-s3-client/tests/common/mod.rs [hunk 5]
```diff
@@ -99,7 +144,8 @@ pub async fn get_mpu_count_for_key(
 #[tokio::test]
 async fn test_sdk_create_object() {
     let sdk_client = get_test_sdk_client().await;
-    let (bucket, prefix) = get_test_bucket_and_prefix("test_sdk_create_object");
+    let bucket = get_test_bucket();
+    let prefix = get_unique_test_prefix("test_sdk_create_object");
```
Comment (passaro) [inline mountpoint-s3-client/tests/common/mod.rs:148]: why the change?

File: mountpoint-s3-client/tests/endpoint_config.rs [hunk 6]
```diff
@@ -36,33 +35,126 @@ async fn run_test<F: FnOnce(&str) -> EndpointConfig>(f: F) {
...
 #[test_case(AddressingStyle::Automatic)]
 #[test_case(AddressingStyle::Path)]
 #[tokio::test]
 async fn test_addressing_style_dualstack_option(addressing_style: AddressingStyle) {
-    run_test(|region| {
-        EndpointConfig::new(region)
-            .addressing_style(addressing_style)
-            .use_dual_stack(true)
-    })
+    let prefix = get_unique_test_prefix("test_dual_stack");
```
Comment (passaro) [inline mountpoint-s3-client/tests/endpoint_config.rs:79]: Maybe we could add the test prefix as a parameter here as well (similarly to `test_addressing_style_region()`)

File: mountpoint-s3-client/tests/endpoint_config.rs [hunk 7]
```diff
@@ -36,33 +35,126 @@ async fn run_test<F: FnOnce(&str) -> EndpointConfig>(f: F) {
...
+async fn test_single_region_access_point(addressing_style: AddressingStyle, arn: bool, prefix: &str) {
+    run_test(
+        |region| EndpointConfig::new(region).addressing_style(addressing_style),
+        &get_unique_test_prefix(prefix),
+        get_test_access_point(arn, AccessPointType::SingleRegion),
+    )
+    .await;
+}
+
+// For Object Labda Access Point, PutObject is not supported,
```
Comment (passaro) [inline mountpoint-s3-client/tests/endpoint_config.rs:120]: typo: Lambda

File: mountpoint-s3-client/tests/endpoint_config.rs [hunk 8]
```diff
@@ -36,33 +35,126 @@ async fn run_test<F: FnOnce(&str) -> EndpointConfig>(f: F) {
...
+    let prefix = get_unique_test_prefix("test_fips_dual_stack");
+    run_test(
+        |region| EndpointConfig::new(region).use_fips(true).use_dual_stack(true),
+        &prefix,
+        get_test_bucket(),
+    )
+    .await;
+}
+
+#[cfg(feature = "access_points_tests")]
```
Comment (passaro) [inline mountpoint-s3-client/tests/endpoint_config.rs:104]: Is "access_points_tests" a new feature? Doesn't it need to be added to Cargo.toml?

File: mountpoint-s3-client/tests/endpoint_config.rs [hunk 9]
```diff
@@ -36,33 +35,126 @@ async fn run_test<F: FnOnce(&str) -> EndpointConfig>(f: F) {
...
 }
 
-#[test_case(AddressingStyle::Automatic)]
-#[test_case(AddressingStyle::Path)]
+#[test_case(AddressingStyle::Automatic, "test_default_addressing_style")]
+#[test_case(AddressingStyle::Path, "test_path_addressing_style")]
 #[tokio::test]
-async fn test_addressing_style_region(addressing_style: AddressingStyle) {
-    run_test(|region| EndpointConfig::new(region).addressing_style(addressing_style)).await;
+async fn test_addressing_style_region(addressing_style: AddressingStyle, prefix: &str) {
```
Comment (passaro) [inline mountpoint-s3-client/tests/endpoint_config.rs:41]: pre-existing: why "region" in `test_addressing_style_region`?

File: mountpoint-s3-client/tests/endpoint_config.rs [hunk 10]
```diff
@@ -36,33 +35,126 @@ async fn run_test<F: FnOnce(&str) -> EndpointConfig>(f: F) {
...
+        |region| EndpointConfig::new(region).addressing_style(addressing_style),
+        &get_unique_test_prefix(prefix),
+        get_test_bucket(),
+    )
+    .await;
 }
 
 #[cfg(feature = "fips_tests")]
 #[tokio::test]
 async fn test_fips_mount_option() {
```
Comment (passaro) [inline mountpoint-s3-client/tests/endpoint_config.rs:52]: pre-existing: why "mount_option" in `test_fips_mount_option`? This is a test of the S3 client, I'd just call this something like: `test_use_fips`, which is the `EndpointConfig` option under test.

File: mountpoint-s3-client/tests/endpoint_config.rs [hunk 11]
```diff
@@ -36,33 +35,126 @@ async fn run_test<F: FnOnce(&str) -> EndpointConfig>(f: F) {
...
+async fn test_addressing_style_region(addressing_style: AddressingStyle, prefix: &str) {
+    run_test(
+        |region| EndpointConfig::new(region).addressing_style(addressing_style),
+        &get_unique_test_prefix(prefix),
+        get_test_bucket(),
+    )
+    .await;
 }
 
 #[cfg(feature = "fips_tests")]
```
Comment (passaro) [inline mountpoint-s3-client/tests/endpoint_config.rs:50]: pre-existing: shouldn't the "fips_tests" feature be in `mountpoint-s3-client`?

File: mountpoint-s3/src/metrics/data.rs (hunk 1 above)
Comment (passaro) [reply mountpoint-s3/src/metrics/data.rs:53]: Try reverting them. `make fmt` does not seem to be the cause to me.

**Review by passaro — CHANGES_REQUESTED** [review]
Just a few comments. In particular about `feature` definitions in Cargo.toml.

Also, I would suggest a rebase+squash: there is a lot of back and forth in the commits which is not particularly relevant, and it would hopefully solve your issues with `make fmt`.

File: mountpoint-s3-client/tests/endpoint_config.rs [hunk 12]
```diff
@@ -36,33 +35,126 @@ async fn run_test<F: FnOnce(&str) -> EndpointConfig>(f: F) {
...
+        &prefix,
+        get_test_bucket(),
+    )
+    .await;
 }
 
-#[test_case(AddressingStyle::Automatic)]
-#[test_case(AddressingStyle::Path)]
+// Transfer acceleration do not work with path style
+#[cfg(feature = "s3_tests")]
```
Comment (passaro) [inline mountpoint-s3-client/tests/endpoint_config.rs:63]: Isn't this redundant? (also on tests below)

**Review by passaro — CHANGES_REQUESTED** [review]
Looking good. Just a minor note.

**Review by passaro — APPROVED** [review]
LGTM
